module_4/gifts/controller/src/main/java/com/epam/esm/controller/CertificateController.java

```diff
@@ -0,0 +1,145 @@
+package com.epam.esm.controller;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Page;
+import org.springframework.data.web.PagedResourcesAssembler;
+import org.springframework.hateoas.PagedModel;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PatchMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+import com.epam.esm.controller.assembler.CertificateViewAssembler;
+import com.epam.esm.controller.converter.CertificateViewConverter;
+import com.epam.esm.controller.view.CertificateView;
+import com.epam.esm.dto.CertificateDto;
+import com.epam.esm.service.CertificateService;
+
+/**
+ * Controller for working with certificates.
+ * 
+ */
+@RestController
+@RequestMapping("/api/v1/certificates")
+public class CertificateController {
+
+	private final CertificateService certificateService;
+	private final CertificateViewConverter certificateConverter;
+	private final PagedResourcesAssembler<CertificateDto> pagedResourcesAssembler;
+	private final CertificateViewAssembler certificateViewAssembler;
+
+	@Autowired
+	public CertificateController(CertificateService certificateService, CertificateViewConverter certificateConverter,
+			PagedResourcesAssembler<CertificateDto> pagedResourcesAssembler,
+			CertificateViewAssembler certificateViewAssembler) {
+		this.certificateService = certificateService;
+		this.certificateConverter = certificateConverter;
+		this.pagedResourcesAssembler = pagedResourcesAssembler;
+		this.certificateViewAssembler = certificateViewAssembler;
+	}
+
+	/**
+	 * Creates and saves passed certificate.
+	 * 
+	 * @param certificateView the certificate to be saved
+	 * @return saved certificate
+	 */
+	@PostMapping
+	@ResponseStatus(HttpStatus.CREATED)
+	@PreAuthorize("hasRole('ADMIN')")
+	public CertificateView create(@RequestBody CertificateView certificateView) {
+		CertificateDto createdCertificateDto = certificateService
+				.create(certificateConverter.convertToDto(certificateView));
```

Comment:
@LizavetaSinitsyna I think that better way to convert data is at the service layer.

Reply:
From my point of view she did correct, because service layer should know nothing about dto from controller

Reply:
I agree, but usually all logic should be executed in the service layer. Dto can also be stored in the service layer. The controller can have access to the DTO from the service.

Reply:
@AliaksandrLiulis, In order to maintain independence between modules I have models on DAO-layer, DTOs on service layer and Views on controller layer. That is why there are two modules where the conversion is implemented: models are converted to dtos and vice versa on service layer, while dtos are converted to views and vice versa on controller layer.